File: keywords.js
import { fetchAllSources } from './dataSources.js';

const AFFIXES = ['ai', 'labs', 'cloud', 'tech'];

/**
 * Normalize company name to domain-safe format
 */
function normalizeName(name) {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]/g, '')
    .substring(0, 25);
}

/**
 * Check if name is valid for domain generation
 */
function isValidName(name) {
  return name.length >= 3 && name.length <= 25;
}

/**
 * Generate domain variations from a base name
 */
function generateVariations(baseName) {
  const variations = new Set();
  const cleanName = normalizeName(baseName);
  
  if (!isValidName(cleanName)) {
    return [];
  }
  
  // Add base name
  variations.add(cleanName);
  
  // Add suffix variations
  AFFIXES.forEach(affix => {
    const withSuffix = `${cleanName}${affix}`;
    if (withSuffix.length <= 25) {
      variations.add(withSuffix);
    }
  });
  
  // Add prefix variations
  AFFIXES.forEach(affix => {
    const withPrefix = `${affix}${cleanName}`;
    if (withPrefix.length <= 25) {
      variations.add(withPrefix);
    }
  });
  
  return Array.from(variations);
}

/**
 * Process all names: deduplicate, normalize, and generate variations
 */
export async function processAllKeywords() {
  console.log('\n=== Processing Keywords ===\n');
  
  // Fetch all sources
  const allNames = await fetchAllSources();
  
  // Deduplicate raw names
  const uniqueNames = Array.from(new Set(allNames));
  console.log(`\nAfter deduplication: ${uniqueNames.length} unique names`);
  
  // Normalize and filter
  const normalizedNames = uniqueNames
    .map(name => normalizeName(name))
    .filter(name => isValidName(name));
  
  const uniqueNormalized = Array.from(new Set(normalizedNames));
  console.log(`After normalization and filtering: ${uniqueNormalized.length} valid names`);
  
  // Generate all variations
  const allVariations = new Set();
  
  for (const name of uniqueNormalized) {
    const variations = generateVariations(name);
    variations.forEach(v => allVariations.add(v));
  }
  
  const result = Array.from(allVariations).sort();
  console.log(`Total variations generated: ${result.length}`);
  
  return result;
}

/**
 * Get batch of keywords for today's run based on day of month
 */
export async function getTodaysBatch(batchSize = 200) {
  console.log('\n=== Determining Today\'s Batch ===\n');
  
  const allKeywords = await processAllKeywords();
  
  // Use day of month (1-31) to rotate through batches
  const now = new Date();
  const dayOfMonth = now.getDate(); // 1-31
  
  const totalBatches = Math.ceil(allKeywords.length / batchSize);
  const currentBatch = (dayOfMonth - 1) % totalBatches;
  
  const startIndex = currentBatch * batchSize;
  const endIndex = Math.min(startIndex + batchSize, allKeywords.length);
  
  const batch = allKeywords.slice(startIndex, endIndex);
  
  console.log(`\n=== Batch Selection ===`);
  console.log(`Total keywords available: ${allKeywords.length}`);
  console.log(`Batch size: ${batchSize}`);
  console.log(`Total batches: ${totalBatches}`);
  console.log(`Day of month: ${dayOfMonth}`);
  console.log(`Current batch: ${currentBatch + 1}/${totalBatches}`);
  console.log(`Range: ${startIndex} to ${endIndex}`);
  console.log(`Keywords in this batch: ${batch.length}`);
  
  return batch;
}

File: keywords.py
import re
from datetime import datetime
from data_sources import fetch_all_sources

AFFIXES = ['ai', 'labs', 'cloud', 'tech']

def normalize_name(name):
    """Normalize company name to domain-safe format"""
    return re.sub(r'[^a-z0-9]', '', name.lower())[:25]

def is_valid_name(name):
    """Check if name is valid for domain generation"""
    return 3 <= len(name) <= 25

def generate_variations(base_name):
    """Generate domain variations from a base name"""
    variations = set()
    clean_name = normalize_name(base_name)
    
    if not is_valid_name(clean_name):
        return []
    
    # Add base name
    variations.add(clean_name)
    
    # Add suffix variations
    for affix in AFFIXES:
        with_suffix = f"{clean_name}{affix}"
        if len(with_suffix) <= 25:
            variations.add(with_suffix)
    
    # Add prefix variations
    for affix in AFFIXES:
        with_prefix = f"{affix}{clean_name}"
        if len(with_prefix) <= 25:
            variations.add(with_prefix)
    
    return list(variations)

def process_all_keywords():
    """Process all names: deduplicate, normalize, and generate variations"""
    print('\n=== Processing Keywords ===\n')
    
    # Fetch all sources
    all_names = fetch_all_sources()
    
    # Deduplicate raw names
    unique_names = list(set(all_names))
    print(f'\nAfter deduplication: {len(unique_names)} unique names')
    
    # Normalize and filter
    normalized_names = [normalize_name(name) for name in unique_names]
    normalized_names = [name for name in normalized_names if is_valid_name(name)]
    
    unique_normalized = list(set(normalized_names))
    print(f'After normalization and filtering: {len(unique_normalized)} valid names')
    
    # Generate all variations
    all_variations = set()
    
    for name in unique_normalized:
        variations = generate_variations(name)
        all_variations.update(variations)
    
    result = sorted(list(all_variations))
    print(f'Total variations generated: {len(result)}')
    
    return result

def get_todays_batch(batch_size=5):
    """Get batch of keywords for today's run based on day of month"""
    print('\n=== Determining Today\'s Batch ===\n')
    
    all_keywords = process_all_keywords()
    
    # Use day of month (1-31) to rotate through batches
    now = datetime.now()
    day_of_month = now.day  # 1-31
    
    total_batches = (len(all_keywords) + batch_size - 1) // batch_size
    current_batch = (day_of_month - 1) % total_batches
    
    start_index = current_batch * batch_size
    end_index = min(start_index + batch_size, len(all_keywords))
    
    batch = all_keywords[start_index:end_index]
    
    print(f'\n=== Batch Selection ===')
    print(f'Total keywords available: {len(all_keywords)}')
    print(f'Batch size: {batch_size}')
    print(f'Total batches: {total_batches}')
    print(f'Day of month: {day_of_month}')
    print(f'Current batch: {current_batch + 1}/{total_batches}')
    print(f'Range: {start_index} to {end_index}')
    print(f'Keywords in this batch: {len(batch)}')
    
    return batch
